### PR TITLE
[v6r20] Escaping strings in DFC: directory level tree

### PR DESCRIPTION
**Pull request by fstagni**
Also minor RSS bug fixes.

### CODE REVIEW

**Comment by fstagni** [comment]
Review, please

File: DataManagementSystem/DB/FileCatalogComponents/DirectoryLevelTree.py [hunk 1]
```diff
@@ -1,124 +1,136 @@
...
-      return result   
+      return result
     if not result['Value']:
       res = S_OK()
       res["DirID"] = 0
       return res
-    
+
     dirID = result['Value']
+    dirID = self.db._escapeString(dirID)
```
Comment (andresailer) [inline DataManagementSystem/DB/FileCatalogComponents/DirectoryLevelTree.py]: I am guessing dirID is an integer before it is passed to escapeString, after which it is a string, which will case an exception in line 84
` ... %d" % dirID `

File: DataManagementSystem/DB/FileCatalogComponents/DirectoryLevelTree.py [hunk 2]
```diff
@@ -1,124 +1,136 @@
...
     """
-    epathString = ','.join( [ 'LPATH%d' % (i+1) for i in range( MAX_LEVELS ) ] )
-    req = 'SELECT LEVEL,%s FROM FC_DirectoryLevelTree WHERE DirID=%d' % (epathString,dirID)
-    result = self.db._query(req,connection)
+    epathString = ','.join(['LPATH%d' % (i + 1) for i in range(MAX_LEVELS)])
+    dirID = self.db._escapeString(dirID)
+    if not dirID['OK']:
+      return dirID
+    dirID = dirID['Value']
+    req = 'SELECT LEVEL,%s FROM FC_DirectoryLevelTree WHERE DirID=%d' % (epathString, dirID)
```
Comment (andresailer) [inline DataManagementSystem/DB/FileCatalogComponents/DirectoryLevelTree.py]: Also a problem when dirID is a string

File: DataManagementSystem/DB/FileCatalogComponents/DirectoryLevelTree.py [hunk 3]
```diff
@@ -158,210 +170,234 @@ def makeDir(self,path):
...
-      resultLock = self.db._query( "START TRANSACTION; ", conn )
-      result = self.db._query(req,conn)
-      req = "UPDATE FC_DirectoryLevelTree SET %s=@tmpvar WHERE DirID=%d; " % (lPath,dirID)   
-      result = self.db._update(req,conn)
-      result = self.db._query( "COMMIT;", conn )
+
+      parentDirID = self.db._escapeString(parentDirID)
+      if not parentDirID['OK']:
+        return parentDirID
+      parentDirID = parentDirID['Value']
```
Comment (andresailer) [inline DataManagementSystem/DB/FileCatalogComponents/DirectoryLevelTree.py]: needs to be integer

File: DataManagementSystem/DB/FileCatalogComponents/DirectoryLevelTree.py [hunk 4]
```diff
@@ -158,210 +170,234 @@ def makeDir(self,path):
...
-    
+
     if dirID == 0:
       return S_ERROR('Root directory ID given')
-    
+
+    dirID = self.db._escapeString(dirID)
+    if not dirID['OK']:
+      return dirID
+    dirID = dirID['Value']
```
Comment (andresailer) [inline DataManagementSystem/DB/FileCatalogComponents/DirectoryLevelTree.py]: integer

File: DataManagementSystem/DB/FileCatalogComponents/DirectoryLevelTree.py [hunk 5]
```diff
@@ -158,210 +170,234 @@ def makeDir(self,path):
...
-  
-  def getDirectoryPath(self,dirID):
+
+  def getDirectoryPath(self, dirID):
     """ Get directory name by directory ID
     """
+    dirID = self.db._escapeString(dirID)
+    if not dirID['OK']:
+      return dirID
+    dirID = dirID['Value']
```
Comment (andresailer) [inline DataManagementSystem/DB/FileCatalogComponents/DirectoryLevelTree.py]: integer

File: DataManagementSystem/DB/FileCatalogComponents/DirectoryLevelTree.py [hunk 6]
```diff
@@ -158,210 +170,234 @@ def makeDir(self,path):
...
+    return S_OK([x[0] for x in result['Value']])
+
+  def getSubdirectoriesByID(self, dirID, requestString=False, includeParent=False):
     """ Get all the subdirectories of the given directory at a given level
     """
 
+    dirID = self.db._escapeString(dirID)
+    if not dirID['OK']:
+      return dirID
+    dirID = dirID['Value']
```
Comment (andresailer) [inline DataManagementSystem/DB/FileCatalogComponents/DirectoryLevelTree.py]: integer

**Comment by chaen** [comment]
Hum... eventhough the proper way to go would be a real library (like SQLAlchemy, even without the ORM part), maybe a quicker and intermediate fix could be to use prepared statement ?

**Comment by atsareg** [comment]
I do not think that integer IDs should be escaped at all. The IDs are the internal DFC objects and are not passed through the client interface

**Comment by fstagni** [comment]
Thanks @andresailer I think I have fixed.

**Comment by andresailer** [comment]
@fstagni  OK, I have not seen anything else now. Thanks

**Pull request merged**